src/skimage/registration/_affine.py

```diff
@@ -0,0 +1,550 @@
+    ndim = reference_image.ndim - 1
+
+    # Initialize with the identity
+    if matrix is None:
+        matrix = np.eye(ndim + 1, dtype=np.float64)
+
+    # Compute the ij grids along each non channel axis
+    grid = np.meshgrid(
+        *[np.arange(n) for n in reference_image.shape[1:]], indexing="ij"
+    )
+
+    # Compute the 1st derivative of the image in each non channel axis
+    grad = [
+        np.gradient(reference_image, axis=k) for k in range(1, reference_image.ndim)
+    ]
+
+    # Vector of gradients eg: [Iy Ix yIy xIy yIy yIx] (see eq. (27))
+    elems = [*grad, *[x * dx for dx, x in product(grad, grid)]]
+
+    # G matrix of eq. (32)
+    g_mat = np.zeros([len(elems)] * 2)
+    for i, j in combinations_with_replacement(range(len(elems)), 2):
+        g_mat[i, j] = g_mat[j, i] = (elems[i] * elems[j] * weights).sum()
+
+    # Model reduction
+    if model.lower() == "translation":
+        h_mat = np.eye(g_mat.shape[0], ndim, dtype=np.float64)
+    elif model.lower() == "euclidean":
+        if ndim == 2:
+            h_mat = np.array(
+                [
+                    [1.0, 0.0, 0.0],
+                    [0.0, 1.0, 0.0],
+                    [0.0, 0.0, 0.0],
+                    [0.0, 0.0, -1.0],
+                    [0.0, 0.0, 1.0],
+                    [0.0, 0.0, 0.0],
+                ],
+            )
+        else:
+            raise NotImplementedError("Eulidean motion model implemented only in 2D")
+    elif model.lower() == "affine":
+        h_mat = np.eye(ndim * (ndim + 1), dtype=matrix.dtype)
+    else:
+        raise NotImplementedError(f"Unsupported motion model {model}")
+
+    for _ in range(max_iter):
+        # Warp each channel
+        moving_image_warp = np.stack(
+            [ndi.affine_transform(plane, matrix) for plane in moving_image]
+        )
+
+        # Compute the error (eq. (26))
+        error_image = reference_image - moving_image_warp
+
+        # Compute b from eq. (33)
+        b = np.array([[(e * error_image * weights).sum()] for e in elems])
+
+        try:
+            # Solve the system
+            r = np.linalg.solve(h_mat.T @ g_mat @ h_mat, h_mat.T @ b)
+            # update the matrix
+            matrix = matrix @ _parameter_vector_to_matrix(r.ravel(), model, ndim)
+
+            if np.linalg.norm(r) < tol:
+                break
+
+        except np.linalg.LinAlgError:
+            warnings.warn("Failed to invert linear system (np.linalg.solve).")
+            break
+
+    return matrix
+
+
+def _param_cost(
+    parameters,
+    reference_image,
+    moving_image,
+    weights,
+    cost,
+    *,
+    model="affine",
+):
+    """
+    Compute the registration cost for the current parameters
+
+    Parameters
+    ----------
+    parameters: ndarray
+        Parameters of the transform.
+    reference_image: ndarray
+        Reference image.
+    moving_image: ndarray
+        Moving image.
+    weights: ndarray | None
```

Comment:
```suggestion
    weights : ndarray | None
```
In the function signature, it doesn't default to `None` though... @jboulanger